File: 05/gpt-5.1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type interval struct {
	start int64
	end   int64
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	ranges := make([]interval, 0)
	available := make([]int64, 0)

	parsingRanges := true
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			parsingRanges = false
			continue
		}

		if parsingRanges {
			var s, e int64
			_, err := fmt.Sscanf(line, "%d-%d", &s, &e)
			if err == nil {
				if e < s {
					s, e = e, s
				}
				ranges = append(ranges, interval{start: s, end: e})
			}
		} else {
			var id int64
			_, err := fmt.Sscanf(line, "%d", &id)
			if err == nil {
				available = append(available, id)
			}
		}
	}

	// Merge overlapping/adjacent ranges
	sort.Slice(ranges, func(i, j int) bool {
		if ranges[i].start == ranges[j].start {
			return ranges[i].end < ranges[j].end
		}
		return ranges[i].start < ranges[j].start
	})

	merged := make([]interval, 0, len(ranges))
	for _, r := range ranges {
		if len(merged) == 0 {
			merged = append(merged, r)
			continue
		}
		last := &merged[len(merged)-1]
		if r.start <= last.end+1 {
			if r.end > last.end {
				last.end = r.end
			}
		} else {
			merged = append(merged, r)
		}
	}

	// Part 1: count available IDs that fall within any merged range
	var part1 int64
	for _, id := range available {
		for _, r := range merged {
			if id < r.start {
				break
			}
			if id <= r.end {
				part1++
				break
			}
		}
	}
	fmt.Printf("Day 05 Part 1: %d\n", part1)

	// Part 2: count all unique IDs covered by any merged range
	var part2 int64
	for _, r := range merged {
		part2 += (r.end - r.start + 1)
	}
	fmt.Printf("Day 05 Part 2: %d\n", part2)
}

File: 05/gpt-5.1/rust/src/main.rs
use std::fs;

#[derive(Debug, Clone, Copy)]
struct Interval {
    start: i64,
    end: i64,
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input");
    let mut lines: Vec<&str> = input
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
        .collect();

    let mut ranges: Vec<Interval> = Vec::new();
    let mut available: Vec<i64> = Vec::new();

    let mut index = 0usize;
    while index < lines.len() && lines[index].contains('-') {
        let line = lines[index];
        let parts: Vec<&str> = line.splitn(2, '-').collect();
        if parts.len() == 2 {
            if let (Ok(mut start), Ok(mut end)) = (parts[0].trim().parse::<i64>(), parts[1].trim().parse::<i64>()) {
                if end < start {
                    std::mem::swap(&mut start, &mut end);
                }
                ranges.push(Interval { start, end });
            }
        }
        index += 1;
    }

    while index < lines.len() {
        let line = lines[index];
        if !line.is_empty() {
            if let Ok(id) = line.parse::<i64>() {
                available.push(id);
            }
        }
        index += 1;
    }

    ranges.sort_by(|a, b| {
        let cmp = a.start.cmp(&b.start);
        if cmp == std::cmp::Ordering::Equal {
            a.end.cmp(&b.end)
        } else {
            cmp
        }
    });

    let mut merged: Vec<Interval> = Vec::new();
    for r in ranges {
        if let Some(last) = merged.last_mut() {
            if r.start <= last.end + 1 {
                if r.end > last.end {
                    last.end = r.end;
                }
                continue;
            }
        }
        merged.push(r);
    }

    let mut part1: i64 = 0;
    for id in &available {
        for r in &merged {
            if *id < r.start {
                break;
            }
            if *id <= r.end {
                part1 += 1;
                break;
            }
        }
    }
    println!("Day 05 Part 1: {}", part1);

    let mut part2: i64 = 0;
    for r in &merged {
        part2 += r.end - r.start + 1;
    }
    println!("Day 05 Part 2: {}", part2);
}